// portal-core/src/main/roleNew/details/RoleDetails.js
require('./roleDetails.scss')

module.exports = {
    template: require('./roleDetails.html'),
    controller: class RoleDetails {
        /* @ngInject */
        constructor($state) {
            this.$state = $state
        }

        $onInit() {
            this.currentUser = this.$state.$current.locals.globals.currentUser

            this.role = this.$state.$current.locals.globals.role

            this.suites = [
                {
                    id: 1,
                    name: 'Suite 1'
                },
                {
                    id: 2,
                    name: 'Suite 2'
                }
            ]

            this.organizations = [
                {
                    id: 1,
                    name: 'Organization 1'
                },
                {
                    id: 2,
                    name: 'Organization 2'
                }
            ]

            this.permissions = [
                {
                    id: 1,
                    name: 'Permission 1'
                },
                {
                    id: 2,
                    name: 'Permission 2'
                }
            ]

            this.role.suites = [
                this.suites[0]
            ]

            this.role.organizations = [
                this.organizations[0]
            ]

            this.role.permissions = [
                this.permissions[0]
            ]

            this.canEditRole = this.currentUser.hasPermission('ROLE_ROLE_WRITE')
        }
    }
}
