// portal-core/src/main/core/config/coreRouteConfig.js
module.exports = /* @ngInject */
    function($stateProvider, routeHelper) {
        $stateProvider

            .state('nav', {
                abstract: true,
                url: '/nav',
                template: '<navigation></navigation>',
                resolve: {
                    //loading this here to ensure page checks login prior to loading
                    currentUser: routeHelper.currentUser
                }
            })

            .state('nav.changelogs', {
                url: '/changelogs',
                template: '<changelogs></changelogs>',
                resolve: {
                    permission: routeHelper.hasPermission('ROLE_CHANGELOG_READ')
                }
            })

            .state('nav.releasenotes', {
                url: '/releasenotes',
                template: '<release-notes></release-notes>'
            })
            
            .state('nav.search', {
                abstract: true,
                template: require('../views/moduleBase.html')
            })

            .state('nav.search.list', {
                url: '/search/:term',
                template: '<search-results></search-results>',
                resolve: {
                    results: function($stateParams, searchService) {
                        return searchService.termSearch($stateParams.term)
                    }
                }
            })
    }
