// portal-core/src/main/core/providers/customActionConfigProvider.js
const _ = require('lodash')

module.exports = 
    /* @ngInject */
    function() {
        let customActionConfig = {}
        
        this.$get = () => {
            let self = {
                customActionConfig: _.cloneDeep(customActionConfig),
                
                getTextButtons: (context, toolbar) => _.get(self.customActionConfig, [context, toolbar, 'textButtons']),
                getIconButtons: (context, toolbar) => _.get(self.customActionConfig, [context, toolbar, 'iconButtons'])
            }
            
            return self
        }
        
        this.addTextButton = (context = 'default', toolbar = 'default', textButton) => {
            if (!_.get(customActionConfig, [context, toolbar, 'textButtons'])) {
                _.set(customActionConfig, [context, toolbar, 'textButtons'], [])
            }

            if (textButton) {
                customActionConfig[context][toolbar]['textButtons'].push(textButton)
            }
        }

        this.addIconButton = (context = 'default', toolbar = 'default', iconButton) => {
            if (!_.get(customActionConfig, [context, toolbar, 'iconButtons'])) {
                _.set(customActionConfig, [context, toolbar, 'iconButtons'], [])
            }

            if (iconButton) {
                customActionConfig[context][toolbar]['iconButtons'].push(iconButton)
            }
        }
    }
