// portal-core/src/main/permissionNew/config/permissionRouteConfig.js
module.exports = /* @ngInject */
    function($stateProvider, routeHelper) {
        $stateProvider
            .state('nav.permission', {
                url: '/permission',
                template: require('../../core/views/moduleBase.html'),
                resolve: {
                    permission: routeHelper.hasPermission('ROLE_PERMISSION_READ')
                }
            })

            .state('nav.permission.add', {
                url: '/add',
                template: '<permission-edit></permission-edit>',
                resolve: {
                    permission: Restangular => {
                        return Restangular.restangularizeElement(null, {adding: true}, 'permissions')
                    }
                }
            })

            .state('nav.permission.details', {
                url: '/details/:key',
                template: '<permission-details></permission-details>',
                resolve: {
                    currentUser: routeHelper.currentUser,
                    permission: routeHelper.get('permissions')
                }
            })

            .state('nav.permission.edit', {
                url: '/edit/:key',
                template: '<permission-edit></permission-edit>',
                resolve: {
                    permission: routeHelper.get('permissions')
                }
            })

            .state('nav.permission.list', {
                url: '/list/',
                template: '<permission-list></permission-list>',
                resolve: {
                    currentUser: routeHelper.currentUser,
                    permissions: (listService, Restangular) => {
                        listService.page = listService.page || 1
                        listService.pageSize = listService.pageSize || 10
                        listService.sort = listService.sort || 'name'

                        let params = {
                            sort: listService.sort
                        }

                        return Restangular.all('permissions').getList(params)
                    }
                }
            })
    }
