// alk-maps/src/test/mocks/MockALKMaps.js
module.exports = function() {
    return {
        Control: {
            Button: sinon.stub(),
            Geolocate: sinon.stub(),
            LayerSwitcher: sinon.stub(),
            MousePosition: sinon.stub(),
            NavPanel: sinon.stub(),
            Navigation: sinon.stub(),
            OverviewMap: sinon.stub(),
            PanZoomBar: sinon.stub(),
            ScaleLine: sinon.stub(),
            SelectFeature: sinon.stub()
        },
        Feature: {
            Vector: sinon.stub()
        },
        Geometry: {
            Point: sinon.stub(),
            LinearRing: sinon.stub(),
            LineString: sinon.stub()
        },
        Geocoder: {
            reverseGeocode: sinon.stub()
        },
        Icon: sinon.stub().returns({
            clone: sinon.stub().returns('clonedIcon')
        }),
        IMAGE: {
            OTHER: {
                clone: sinon.stub()
            }
        },
        Layer: {
            BaseMap: sinon.stub(),
            HeatMap: sinon.stub(),
            Markers: sinon.stub(),
            Traffic: sinon.stub(),
            WeatherRadar: sinon.stub(),
            WeatherAlerts: sinon.stub(),
            Vector: sinon.stub(),
            VectorMarkers: sinon.stub()
        },
        LonLat: sinon.stub().returns({ transform: sinon.stub()}),
        Map: sinon.stub(),
        Marker: sinon.stub(),
        Popup: {
            FramedCloud: sinon.stub()
        },
        Projection: sinon.stub(),
        Renderer: {
            symbol: {
                truck: [1, 2, 3]
            }
        },
        Size: sinon.stub(),
        STYLE: {
            DEFAULT: 123,
            Blah: 234
        },
        Strategy: {
            Cluster: sinon.stub()
        },
        StyleMap: sinon.stub()
    }
}
