// portal-core/src/main/role/components/RoleList.js
const _ = require('lodash')
require('../styles/roleList.scss')

module.exports = {
    template: require('../views/roleList.html'),
    controller: class RoleList {
        /* @ngInject */
        constructor($filter, $state, listService) {
            this.$filter = $filter
            this.$state = $state
            this.listService = listService
        }

        $onInit() {
            //TODO: replace this workaround when new router is available
            this.roles = this.$state.$current.locals.globals.roles
            this.currentUser = this.$state.$current.locals.globals.currentUser

            this.staticResults = _.map(this.roles, role => ({
                icon: 'people_outline',
                name: role.name,
                type: 'static',
                key: role.key,
                term: role.key,
                param: 'app'
            }))

            this.currentState = this.$state.current

            this.total = this.roles.length

            this.context = 'roleList'

            this.toolbarTitle = 'role.roleList.title'

            this.cols = [
                {
                    name: 'role.roleList.name',
                    template: `<a ng-if="$ctrl.currentUser.hasPermission('ROLE_ROLE_WRITE')" ng-href="/#/nav/role/edit/{{row.id}}" class="list-row-link" ng-bind="row.name"></a>
                               <span ng-if="!$ctrl.currentUser.hasPermission('ROLE_ROLE_WRITE')" ng-bind="row.name"></span>`
                },
                {
                    name: 'role.roleList.description',
                    path: 'description'
                },
                {
                    name: 'role.roleList.permissions',
                    template: "<span>{{'role.roleList.permission_count_label' | translate: {result: row.permissions.length} }} </span>"
                },
                {
                    name: 'role.roleList.org_types',
                    template: `<div ng-if="row.organizationTypes.length < 3" ng-repeat="organizationType in row.organizationTypes" ng-bind="organizationType.displayName"></div>
                               <div ng-if="row.organizationTypes.length > 2 || row.organizationTypes.length === 0" title="{{$ctrl.createLargeTitle(row, 'organizationTypes', 'displayName')}}">
                                   {{'role.roleList.organization_count_label'| translate: {result: row.organizationTypes.length} }}
                               </div>`
                },
                {
                    name: 'role.roleList.applications',
                    template: `<div ng-if="row.applications.length < 3" ng-repeat="application in row.applications" ng-bind="application.clientId"> </div>
                               <div ng-if="row.applications.length > 2 || row.applications.length === 0" title="{{$ctrl.createLargeTitle(row, 'applications', 'name')}}">
                                  {{'role.roleList.application_count_label'| translate: {result: row.applications.length} }}
                               </div>`
                }
            ]

            this.rows = this.roles

            this.toolbarOptions = {
                densityButton: true,
                filterButton: true
            }

            this.tableOptions = {
                rowSelect: true
            }

            this.paginationOptions = {
                boundaryLinks: true,
                pageSelect: true,
                pageSizes: [10, 25, 50]
            }

            this.chipFilterOptions = {
                entities: this.entities,
                entityChips: this.entityChips,
                staticResults: this.staticResults
            }

            this.cardContent = '<role-card role="row" current-user="$ctrl.currentUser"></role-card>'

            this.params = {
                page: this.listService.page - 1,
                pageSize: this.listService.pageSize
            }

            this.getRows()
        }

        getRows() {
            let start = this.params.page * this.params.pageSize
            let end = start + this.params.pageSize
            this.rows = _.slice(this.roles, start, end)
        }

        changePage(params) {
            this.params.page = params.page
            this.params.pageSize = params.pageSize

            this.listService.page = params.page + 1
            this.listService.pageSize = params.pageSize

            return this.getRows()
        }

        changeDensity(params) {
            return this.listService.density = params.density
        }

        chipsUpdated(selectedChips) {
            if (_.get(selectedChips.selectedChips, 'length')) {
                let results = _.reduce(selectedChips.selectedChips, (result, chip) => {
                    let found = this.$filter('filter')(this.roles, chip.term)
                    result.push(found)
                    return result
                }, [])

                //find the results that match all chips and put them first in the results
                //then put the rest of the results in the list - this should relatively
                //simulate how elastic search returns search results
                this.rows = _.uniq(_.intersection(...results).concat(_.flatten(results)))

                this.total = _.size(this.rows)
            } else {
                this.getRows()

                this.total = _.size(this.roles)
            }

            //stop the state reload from happening
            return true
        }
    }
}
