// portal-core/src/main/core/components/CandyBar.js
const _ = require('lodash')
require('../styles/candyBar.scss')

module.exports = {
    template: require('../views/candyBar.html'),
    controller: class CandyBar {
        /* @ngInject */
        constructor(candyBarConfig, themeService) {
            this.candyBarConfig = candyBarConfig
            this.themeService = themeService
        }

        $onInit() {
            this.items = _.map(this.candyBarConfig.items, item => {
                let newItem = _.cloneDeep(item)
                newItem.colspan = item.colspan || 3
                newItem.rowspan = item.rowspan || 1
                return newItem
            })

            this.themes = _.map(this.candyBarConfig.themes, theme => {
                let newTheme = _.cloneDeep(theme)
                newTheme.colspan = theme.colspan || 3
                newTheme.rowspan = theme.rowspan || 1
                return newTheme
            })

            this.children = this.items.length > 0 || this.themes.length > 0

            this.setTheme = theme => this.themeService.setTheme(theme)
        }
    }
}
