// portal-core/src/main/core/components/Changelogs.js
require('../styles/changelogs.scss')

module.exports = {
    template: require('../views/changelogs.html'),
    controller: class Changelogs {
        /* @ngInject */
        constructor($log, $showdown, $q, staticContentService) {
            this.$log = $log
            this.$showdown = $showdown
            this.$q = $q
            this.staticContentService = staticContentService
        }

        $onInit() {
            return this.staticContentService.getContent('modules.json')
                .then((modules=[]) => {
                    let promises = modules
                        .concat(['portal-core'])
                        .map(module => {
                            let file = `changelogs/${module}.md`
                            return this.staticContentService.getContent(file)
                                .then(content => ({
                                    name: module,
                                    content: this.parseMarkdown(content)
                                }))
                                .catch(err => {
                                    this.$log.error('Unable to load changelog: ', err)
                                    return {
                                        name: module,
                                        content: 'Change log unavailable.'
                                    }
                                })
                        })
                    return this.$q.all(promises)
                }).then(changelogs => this.changelogs = changelogs)
        }

        parseMarkdown(input) {
            let html = this.$showdown.makeHtml(input)

            //add target="blank" to links
            html = html.replace(/(<a\s)([^>]+>)/g, '$1target="blank" $2')

            return html
        }
    }
}
