// portal-core/src/main/core/components/DensityToggle.js
require('../styles/densityToggle.scss')

module.exports = {
    template: require('../views/densityToggle.html'),
    bindings: {
        onChange: '&'
    },
    controller: class DensityToggle {
        /* @ngInject */
        constructor($analytics, $state, $stateParams, $mdMedia) {
            this.$analytics = $analytics
            this.$state = $state
            this.$stateParams = $stateParams
            this.$mdMedia = $mdMedia
        }

        $onInit() {
            let defaultView = this.$mdMedia('xs') ? 'low' : 'high'

            if (this.$stateParams.density === 'high' || this.$stateParams.density === 'low') {
                this.density = this.$stateParams.density
            } else {
                this.density = defaultView
            }

            this.onChange({value: this.density})
        }

        click(value) {
            let params = this.$stateParams
            this.density = value
            params.density = this.density

            this.$analytics.eventTrack(this.$state.current.name, {category: 'Density View', label: this.density})
            this.$state.go(this.$state.current, params)
        }
    }
}
