// portal-core/src/main/core/components/MapView.js
const _ = require('lodash')
require('../styles/mapView.scss')

module.exports = {
    template: require('../views/mapView.html'),
    bindings: {
        mapSettings: '<',
        mapLoaded: '&'
    },
    controller: class MapView {
        /* @ngInject */
        constructor($analytics, $log, $timeout, mapApiKey, pnetMap) {
            this.$analytics = $analytics
            this.$log = $log
            this.$timeout = $timeout
            this.mapApiKey = mapApiKey
            this.pnetMap = pnetMap
        }

        $onInit() {
            //create the alk map
            return this.$timeout(() => this.loadMap())
                .then(() => this.mapLoaded({map: this.map}))
        }

        $onDestroy() {
            try {
                _.result(this.map, 'destroy')
            } catch (err) {
                this.$log.trace('Error destroying map', err)
            }
        }

        loadMap() {
            this.map = this.pnetMap.init(this.mapApiKey, 'map', this.mapSettings || {minimalist: true})
        }

        zoomIn() {
            this.map.getMap().zoomIn()
            this.$analytics.eventTrack('Zoom', {category: 'Map Controls', label: 'In'})
        }

        zoomOut() {
            this.map.getMap().zoomOut()
            this.$analytics.eventTrack('Zoom', {category: 'Map Controls', label: 'Out'})
        }
    }
}
