// portal-core/src/main/core/components/PageHeader.js
const _ = require('lodash')
require('../styles/pageHeader.scss')

module.exports = {
    bindings: {
        title: '@',
        subtitle: '@',
        backButtonHandler: '<'
    },
    template: require('../views/pageHeader.html'),
    controller: class PageHeader {
        /* @ngInject */
        constructor($state, $window, titleService) {
            this.$state = $state
            this.$window = $window
            this.titleService = titleService
        }

        $onInit() {
            //defer to let titleService load
            _.defer(() => this.titleService.setTitle(this.title))
        }
        
        $onDestroy() {
            this.titleService.resetTitle()
        }

        handleBack() {
            this.transitioning = true
            if (_.isString(this.backButtonHandler)) {
                this.$state.go(this.backButtonHandler)
            } else if (_.isFunction(this.backButtonHandler)) {
                this.backButtonHandler()
            } else if (_.isObject(this.backButtonHandler)) {
                this.$state.go(this.backButtonHandler.state, this.backButtonHandler.params)
            } else {
                this.$window.history.back()
            }
        }
    }
}
