// portal-core/src/main/application/components/ApplicationList.js
const _ = require('lodash')
require('../styles/applicationList.scss')

module.exports = {
    template: require('../views/applicationList.html'),
    controller: class ApplicationList {
        /* @ngInject */
        constructor($filter, $state, listService) {
            this.$filter = $filter
            this.$state = $state
            this.listService = listService
        }

        $onInit() {
            //TODO: replace this workaround when new router is available
            this.applications = this.$state.$current.locals.globals.applications
            this.currentUser = this.$state.$current.locals.globals.currentUser

            this.staticResults = _.map(this.applications, app => ({
                icon: 'laptop',
                name: app.name,
                type: 'static',
                key: app.key,
                term: app.key,
                param: 'app'
            }))

            this.currentState = this.$state.current

            this.total = this.applications.length

            this.context = 'applicationList'

            this.toolbarTitle = 'application.common.applications'

            this.cols = [
                {
                    name: 'application.common.name',
                    template: `<a ng-if="$ctrl.currentUser.hasPermission('ROLE_APPLICATION_WRITE')" ng-href="/#/nav/application/edit/{{row.id}}" class="list-row-link" ng-bind="row.name"></a>
                               <span ng-if="!$ctrl.currentUser.hasPermission('ROLE_APPLICATION_WRITE')" ng-bind="row.name"></span>`
                },
                {
                    name: 'application.common.id',
                    path: 'clientId'
                },
                {
                    name: 'application.common.description',
                    path: 'description'
                },
                {
                    name: 'application.common.realm',
                    template: `<span ng-bind="row.realms.join(', ')"></span>`
                }
            ]

            this.rows = this.applications

            this.toolbarOptions = {
                densityButton: true,
                filterButton: true
            }

            this.tableOptions = {
                rowSelect: true
            }

            this.paginationOptions = {
                boundaryLinks: true,
                pageSelect: true,
                pageSizes: [10, 25, 50]
            }

            this.chipFilterOptions = {
                entities: this.entities,
                entityChips: this.entityChips,
                staticResults: this.staticResults
            }

            this.cardContent = '<application-card application="row" current-user="$ctrl.currentUser"></application-card>'

            this.params = {
                page: this.listService.page - 1,
                pageSize: this.listService.pageSize
            }

            this.getRows()
        }

        getRows() {
            let start = this.params.page * this.params.pageSize
            let end = start + this.params.pageSize
            this.rows = _.slice(this.applications, start, end)
        }

        changePage(params) {
            this.params.page = params.page
            this.params.pageSize = params.pageSize

            this.listService.page = params.page + 1
            this.listService.pageSize = params.pageSize

            return this.getRows()
        }

        changeDensity(params) {
            return this.listService.density = params.density
        }

        chipsUpdated(selectedChips) {
            if (_.get(selectedChips.selectedChips, 'length')) {
                let results = _.reduce(selectedChips.selectedChips, (result, chip) => {
                    let found = this.$filter('filter')(this.applications, chip.term)
                    result.push(found)
                    return result
                }, [])

                //find the results that match all chips and put them first in the results
                //then put the rest of the results in the list - this should relatively
                //simulate how elastic search returns search results
                this.rows = _.uniq(_.intersection(...results).concat(_.flatten(results)))

                this.total = _.size(this.rows)
            } else {
                this.getRows()

                this.total = _.size(this.applications)
            }

            //stop the state reload from happening
            return true
        }
    }
}
