// portal-core/src/gulp/webpack.config.js
// const webpack = require('webpack')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const path = require('path')

module.exports = {
    entry: {
        app: [
            './src/main/app.js'
        ]
    },
    output: {
        path: `${__dirname}/dist`,
        filename: 'bundle.js'
    },
    module: {
        loaders: [
            {
                test: /\.js$/,
                exclude: /node_modules\/(?!portal-core|.+portal-modules)/,
                loaders: ['ng-annotate?map=false', 'babel']
            },
            {test: /\.css$/, loaders: ['style-loader', 'css-loader']},
            {test: /\.json$/, loaders: ['json-loader']},
            {test: /\.scss$/, loaders: ['style-loader', 'css-loader', 'postcss-loader']},
            {test: /\.html$/, loader: 'raw-loader'},
            {test: /\.(png|svg|gif|jpg)/, loader: 'file?name=images/[name].[ext]'}
        ]
    },
    resolveLoader: {
        modulesDirectories: [
            path.resolve('./node_modules'),
            path.resolve('./node_modules/portal-core/node_modules')
        ]
    },
    plugins: [
        new HtmlWebpackPlugin({
            template: path.resolve('./node_modules/portal-core/src/main/core/static/index.html')
        })
    ],
    postcss: () => ([
        //scss syntax
        require('precss')
    ])
}
