// portal-core/src/main/loginNew/providers/urlConfigProvider.js
var _ = require('lodash')

module.exports = /* @ngInject */ function() {
    var urlConfig = {
        envs: [
            {name: 'dev', matcher: /^(localhost|192.|10.)/},
            {name: 'dev', matcher: /dev/},
            {name: 'qa', matcher: /qa/},
            {name: 'staging', matcher: /staging/},
            {name: 'prod', matcher: /canary/},
            {name: 'prod', matcher: /fleethealth/}
        ],

        urls: {
            local: 'http://localhost:8080',
            dev: 'http://security-gateway-rp-dev.connectedfleet.io:8080',
            qa: 'http://security-gateway-rp-qa.connectedfleet.io:8080',
            staging: 'https://security-gateway-rp-staging.connectedfleet.io',
            prod: 'https://security-gateway-rp.fleethealth.io'
        }
    }

    this.$get = /* @ngInject */ function($location) {
        var api = {
            getEnv: function() {
                var host = $location.host()
                return _.find(urlConfig.envs, function(env) {
                    return env.matcher.test(host)
                })
            },

            getUrl: function() {
                var env = api.getEnv() || {}
                return urlConfig.urls[env.name] || urlConfig.urls.dev
            }
        }

        return api
    }

    this.addEnv = function(name, matcher) {
        urlConfig.envs.push({
            name: name,
            matcher: matcher
        })
    }

    this.addUrl = function(env, url) {
        urlConfig.urls[env] = url
    }
}
