// portal-core/src/main/core/list/ListService.js
const _ = require('lodash')

module.exports = class ListService {
    /* @ngInject */
    constructor($analytics, $mdMedia, $state) {
        this.$analytics = $analytics
        this.$mdMedia = $mdMedia
        this.$state = $state

        this.init = this.onInit()
    }

    onInit() {
        this.showFilter = false
        this.selectedRows = []
        this.page = 1
        this.pageSize = 10
        this.sort = ''
        this.density = this.$mdMedia('xs') ? 'low' : 'high'

        this.page = 1
        this.pageSize = 10
    }

    onDestroy() {
        return this.onInit()
    }
    
    getSelectedRows(rows) {
        return _.filter(rows, row => !_.isNil(_.find(this.selectedRows, selectedRow => selectedRow.key === row.key)))
    }

    selectRow(row) {
        this.selectedRows.push(row)
    }

    deselectRow(row) {
        let _row = _.cloneDeep(row)

        delete _row.$$hashKey

        return _.remove(this.selectedRows, selectedRow => {
            let _selectedRow = _.cloneDeep(selectedRow)

            delete _selectedRow.$$hashKey

            return _.isEqual(_row, _selectedRow)
        })
    }
}
