// portal-core/src/main/permissionNew/edit/PermissionEdit.js
require('./permissionEdit.scss')

module.exports = {
    template: require('./permissionEdit.html'),
    controller: class PermissionEdit {
        /* @ngInject */
        constructor($log, $state, $translate, loadingService, toastService) {
            this.$log = $log
            this.$state = $state
            this.$translate = $translate
            this.loadingService = loadingService
            this.toastService = toastService
        }

        $onInit() {
            this.permission = this.$state.$current.locals.globals.permission

            return this.loadTranslations()
        }

        loadTranslations() {
            let translations = [
                'permissionNew.permissionEdit.error.on_create',
                'permissionNew.permissionEdit.error.on_update',
                'permissionNew.permissionEdit.success.on_create',
                'permissionNew.permissionEdit.success.on_update'
            ]
            return this.$translate(translations)
                .then(res => this.res = res)
        }

        save() {
            let event = this.permission.adding ? 'on_create' : 'on_update'
            this.loadingService.loading(true)
            return this.permission.put()
                .then(() => {
                    this.toastService.show(this.res[`permissionNew.permissionEdit.success.${event}`])
                    return this.$state.go('nav.permission.list')
                })
                .catch(err => {
                    this.loadingService.loading(false)
                    this.$log.error('Error saving permission.', err)
                    return this.toastService.show(this.res[`permissionNew.permissionEdit.error.${event}`], {theme: 'warn'})
                })
                .finally(() => this.loadingService.loading(false))
        }
    }
}
