// portal-core/src/main/roleNew/config/roleRouteConfig.js
module.exports = /* @ngInject */
    function($stateProvider, routeHelper) {
        $stateProvider
            .state('nav.role', {
                url: '/role',
                template: require('../../core/views/moduleBase.html'),
                resolve: {
                    permission: routeHelper.hasPermission('ROLE_ROLE_READ')
                }
            })

            .state('nav.role.add', {
                url: '/add',
                template: '<role-edit></role-edit>',
                resolve: {
                    applications: Restangular => {
                        return Restangular.all('applications').getList()
                    },
                    organizations: Restangular => {
                        return Restangular.all('organizationTypes').all('realms').all('OEM').getList()
                    },
                    permissions: Restangular => {
                        return Restangular.all('permissions').getList()
                    },
                    role: Restangular => {
                        return Restangular.restangularizeElement(null, {adding: true}, 'roles')
                    }
                }
            })

            .state('nav.role.details', {
                url: '/details/:key',
                template: '<role-details></role-details>',
                resolve: {
                    currentUser: routeHelper.currentUser,
                    permissions: Restangular => {
                        return Restangular.all('permissions').getList()
                    },
                    role: routeHelper.get('roles')
                }
            })

            .state('nav.role.edit', {
                url: '/edit/:key',
                template: '<role-edit></role-edit>',
                resolve: {
                    applications: Restangular => {
                        return Restangular.all('applications').getList()
                    },
                    organizations: Restangular => {
                        return Restangular.all('organizationTypes').all('realms').all('OEM').getList()
                    },
                    permissions: Restangular => {
                        return Restangular.all('permissions').getList()
                    },
                    role: routeHelper.get('roles')
                }
            })

            .state('nav.role.list', {
                url: '/list/',
                template: '<role-list></role-list>',
                resolve: {
                    currentUser: routeHelper.currentUser,
                    roles: (listService, Restangular) => {
                        listService.page = listService.page || 1
                        listService.pageSize = listService.pageSize || 10
                        listService.sort = listService.sort || 'name'

                        let params = {
                            sort: listService.sort
                        }

                        return Restangular.all('roles').getList(params)
                    }
                }
            })
    }
