// portal-core/src/main/login/components/ForgotPassword.js
var _ = require('lodash')

module.exports = {
    template: require('../views/forgotPassword.html'),
    controller: class ForgotPassword {
        /* @ngInject */
        constructor($log, $state, $translate, toastService, Restangular) {
            this.$log = $log
            this.$state = $state
            this.$translate = $translate
            this.toastService = toastService
            this.Restangular = Restangular
        }

        $onInit() {
            this.user = {}

            this.$translate([
                'login.common.err_unexpected',
                'login.reset.confirmation_title'
            ]).then(translations => {
                this.res = translations
            })
        }

        initiateResetPassword() {
            return this.Restangular.all('users').all('mailForgotPasswordLink').post(this.user)
                .then(response => {
                    if (response.success) {
                        this.toastService.show(this.res['login.reset.confirmation_title'], {hideDelay: 0})
                        return this.$state.go('auth.login')
                    } else {
                        this.$log.error('Error mailing password reset link', response)
                        var msg = _.get(response, 'errorMessage', this.res['login.common.err_unexpected'])
                        this.toastService.show(msg, {theme: 'warn'})
                    }
                })
                .catch(err => {
                    this.$log.error('Error mailing password reset link', err)
                    this.toastService.show(this.res['login.common.err_unexpected'], {theme: 'warn'})
                })
        }
    }
}
