// portal-core/src/main/login/components/UserEditSecurity.js
module.exports = {
    template: require('../views/userEditSecurity.html'),
    controller: class UserEditSecurity {
        /* @ngInject */
        constructor($state) {
            this.$state = $state
        }

        $onInit() {
            //TODO: replace this workaround when new router is available
            this.currentUser = this.$state.$current.locals.globals.currentUser
            this.user = this.$state.$current.locals.globals.user

            this.editingSelf = this.user.id === this.currentUser.uid
            this.clearPasswordFields()
        }

        clearPasswordFields() {
            this.currentPassword = ''
            this.newPassword = ''
            this.confirmPassword = ''
        }

        // Validates that 'New password' and 'Confirm password' match
        comparePasswordFields() {
            //TODO:fix the messages, it is not working correctly
            // new password must be different
            this.passwordForm.newPassword.$setValidity(
                'different',
                this.currentPassword !== this.newPassword
            )

            // Confirm and new password must match
            this.passwordForm.confirmPassword.$setValidity(
                'fieldEquality',
                this.newPassword === this.confirmPassword
            )
        }
    }
}
