// alk-maps/src/main/layers/styles/LandmarkLayerStyle.js
module.exports = function(ALKMaps, options) {
    ALKMaps.Renderer.symbol.pin = [15, 59, 30, 23, 30, 21, 31, 20, 31, 11, 30, 10, 30, 9, 29, 8, 29, 7, 28, 6, 27, 5,
        26, 4, 25, 3, 24, 2, 22, 1, 20, 0, 11, 0, 9, 1, 7, 2, 6, 3, 5, 4, 4, 5, 3, 6, 2, 7, 2, 8, 1, 9, 1, 10, 0, 11,
        0, 20, 1, 21, 1, 23, 14, 59, 15, 59]

    return new ALKMaps.Style({
        strokeColor: '#808080',
        pointRadius: '${radius}',
        strokeWidth: 1,
        fillColor: '${fill}',
        fontSize: '11px',
        fontColor: '#116666',
        label: '${label}',
        graphicName: '${graphic}',
        graphicYOffset: -10,
        labelYOffset: '${labelOffset}',
        fillOpacity: 0.7,
        cursor: 'pointer'
    }, {
        context: {
            fill: function(feature) {
                return feature.colorCode || options.colorCode || '#FFE51E'
            },
            radius: function(feature) {
                return feature.cluster ? Math.sqrt(feature.cluster.length / Math.PI) * 8 : 8
            },
            graphic: function(feature) {
                return feature.cluster ? 'circle' : 'pin'
            },
            label: function(feature) {
                return feature.cluster ? feature.cluster.length : ''
            },
            labelOffset: function(feature) {
                return feature.cluster ? 0 : -12
            }
        }
    })
}
