// portal-core/src/main/core/directives/sortableHeaderDirective.js
var _ = require('lodash')

module.exports = function() {
    return {
        restrict: 'A',
        template: require('../views/sortableHeader.html'),
        transclude: true,
        controllerAs: '$ctrl',
        scope: {
            sortableHeader: '@'
        },
        controller: class SortableHeader {
            /* @ngInject */
            constructor($stateParams, $state, $scope) {
                this.$stateParams = $stateParams
                this.$state = $state
                this.$scope = $scope
            }

            $onInit() {
                this.currentSort = this.$stateParams.sort

                if (_.startsWith(this.currentSort, '-')) {
                    this.sortDirection = '-'
                    this.currentSort = this.currentSort.slice(1)
                } else {
                    this.sortDirection = '+'
                }
            }

            performSort() {
                let newSort = this.$scope.sortableHeader

                if (this.currentSort === newSort && this.sortDirection === '+') {
                    // Toggle current sort to Desc
                    newSort = `-${newSort}`
                }

                this.$state.go(this.$state.current.name, {sort:newSort}, {reload: true})
            }
        }
    }
}
