// portal-core/src/main/organization/listDialog/OrganizationListDialog.js
const _ = require('lodash')

require('./organizationListDialog.scss')

module.exports = {
    template: require('./organizationListDialog.html'),
    clickOutsideToClose: true,
    escapeToClose: true,
    controllerAs: '$ctrl',
    controller: class OrganizationListDialog {
        /* @ngInject */
        constructor($filter, $mdDialog, $translate, assignee, listService, organizations) {
            this.$filter = $filter
            this.$mdDialog = $mdDialog
            this.$translate = $translate
            this.assignee = assignee
            this.listService = listService
            this.organizations = organizations

            this.init = this.onInit()
        }

        onInit() {
            this.context = 'organizationListDialog'

            this.toolbarTitle = 'organization.organizationList.title'

            this.cols = [
                {
                    name: 'organization.organizationList.cols.name',
                    path: 'name',
                    sortable: true
                },
                {
                    name: 'organization.organizationList.cols.actions',
                    template: '<modal-action-icons row="row"></modal-action-icons>'
                }
            ]

            this.staticResults = _.map(this.organizations, organization => ({
                icon: 'person',
                name: organization.name,
                type: 'static',
                key: organization.id,
                term: organization.name,
                param: 'organization'
            }))

            this.rows = this.organizations

            this.total = _.size(this.rows)

            this.toolbarOptions = {
                filterButton: true
            }

            this.tableOptions = {
                rowSelect: true
            }

            this.paginationOptions = {
                boundaryLinks: true,
                pageSizes: [10, 25, 50]
            }

            this.chipFilterOptions = {
                staticResults: this.staticResults
            }

            this.listService.page = 1
            this.listService.pageSize = 10

            this.params = {
                page: this.listService.page - 1,
                pageSize: this.listService.pageSize
            }

            this.getRows()

            return this.loadTranslations()
        }

        loadTranslations() {
            let translations = [
                'organization.organizationListDialog.subheader',
                'organization.organizationListDialog.instructions'
            ]

            return this.$translate(translations)
                .then(res => this.res = res)
        }

        getRows() {
            let start = this.params.page * this.params.pageSize
            let end = start + this.params.pageSize
            this.rows = _.slice(this.organizations, start, end)
        }

        changePage(params) {
            this.params.page = params.page
            this.params.pageSize = params.pageSize

            this.listService.page = params.page + 1
            this.listService.pageSize = params.pageSize

            return this.getRows()
        }

        checkChipsCustom() {
            return !!this.chips
        }

        clearChipsCustom() {
            this.getRows()
            return _.remove(this.chips)
        }

        chipsUpdated(params) {
            if (_.get(params.selectedChips, 'length')) {
                let results = _.reduce(params.selectedChips, (result, chip) => {
                    let found = this.$filter('filter')(this.organizations, chip.term)
                    result.push(found)
                    return result
                }, [])

                //find the results that match all chips and put them first in the results
                //then put the rest of the results in the list - this should relatively
                //simulate how elastic search returns search results
                this.rows = _.uniq(_.intersection(...results).concat(_.flatten(results)))

                this.total = _.size(this.rows)
            } else {
                this.getRows()

                this.total = _.size(this.organizations)
            }

            //stop the state reload from happening
            return true
        }

        cancel() {
            this.$mdDialog.cancel()
        }

        save() {
            let organizations = _.filter(this.organizations, {active: true})
            this.$mdDialog.hide(organizations)
        }
    }
}
