// portal-core/src/main/permission/config/permissionRouteConfig.js
module.exports = /* @ngInject */
    function($stateProvider, routeHelper) {

        //add to routes
        $stateProvider
            .state('nav.permission', {
                url: '/permission',
                template: require('../../core/views/moduleBase.html'),
                resolve: {
                    permission: routeHelper.hasPermission('ROLE_PERMISSION_READ')
                }
            })

            .state('nav.permission.list', {
                url: '/list/?density',
                template: '<permission-list></permission-list>',
                resolve: {
                    currentUser: routeHelper.currentUser,
                    permissions: function(listService, Restangular) {
                        listService.page = listService.page || 1
                        listService.pageSize = listService.pageSize || 10

                        return Restangular.all('permissions').getList({sort: 'name'})
                    }
                }
            })

            .state('nav.permission.edit', {
                url: '/edit/:key',
                template: '<permission-edit></permission-edit>',
                resolve: {
                    permission: routeHelper.hasPermission('ROLE_PERMISSION_WRITE'),
                    perm: routeHelper.get('permissions')
                }
            })

            .state('nav.permission.add', {
                url: '/add',
                template: '<permission-edit></permission-edit>',
                resolve: {
                    permission: routeHelper.hasPermission('ROLE_PERMISSION_CREATE'),
                    perm: Restangular => Restangular.restangularizeElement(null, {
                        adding: true
                    }, 'permissions')
                }
            })
    }
