// portal-core/src/main/userNew/details/profile/UserDetailsProfile.js
require('./userDetailsProfile.scss')

module.exports = {
    template: require('./userDetailsProfile.html'),
    controller: class UserDetails {
        /* @ngInject */
        constructor($state) {
            this.$state = $state
        }

        $onInit() {
            //TODO: replace this workaround when new router is available
            this.user = this.$state.$current.locals.globals.user
            this.currentUser = this.$state.$current.locals.globals.currentUser

            this.canEditUser = this.currentUser.canEditUser(this.user)

            //populate some fake data for now
            this.user.userOrganizations.push({
                organizationId: 'OEM',
                organizationTypeId: 'OEM',
                organizationName: 'OEM'
            })
            this.user.userOrganizations[0].roles = [
                {
                    roleId: 123,
                    roleName: 'Super Admin'
                },
                {
                    roleId: 567,
                    roleName: 'Not so Super Admin'
                }
            ]
            this.user.userOrganizations[0].permissions = [
                {
                    permissionId: 489,
                    permissionName: 'A Permission'
                },
                {
                    permissionId: 556,
                    permissionName: 'Read Permissions'
                }
            ]

            this.user.userOrganizations[1].roles = [
                {
                    roleId: 145,
                    roleName: 'Some Role'
                }
            ]
            this.user.userOrganizations[1].permissions = [
                {
                    permissionId: 529,
                    permissionName: 'Some Permission'
                }
            ]
        }
    }
}
