// portal-core/src/main/permissionNew/list/PermissionList.js
const _ = require('lodash')

require('./permissionList.scss')

module.exports = {
    template: require('./permissionList.html'),
    controller: class PermissionList {
        /* @ngInject */
        constructor($filter, $state, listService) {
            this.$filter = $filter
            this.$state = $state
            this.listService = listService
        }

        $onInit() {
            //TODO: replace this workaround when new router is available
            this.permissions = this.$state.$current.locals.globals.permissions
            this.currentUser = this.$state.$current.locals.globals.currentUser

            this.staticResults = _.map(this.permissions, permission => ({
                icon: 'lock_outline',
                name: permission.name,
                type: 'static',
                key: permission.key,
                term: permission.key,
                param: 'app'
            }))

            this.currentState = this.$state.current

            this.total = this.permissions.length

            this.context = 'permissionList'

            this.toolbarTitle = 'permissionNew.permissionList.title'

            this.cols = [
                {
                    name: 'permissionNew.permissionList.cols.name',
                    template: `<a ng-href="/#/nav/permission/details/{{row.id}}" 
                                  ng-if="$ctrl.currentUser.hasPermission('ROLE_PERMISSION_READ')"
                                  class="list-row-link" 
                                  ng-bind="row.name"></a>
                               <span ng-if="!$ctrl.currentUser.hasPermission('ROLE_PERMISSION_READ')"
                                     ng-bind="row.name"></span>`
                },
                {
                    name: 'permissionNew.permissionList.cols.description',
                    path: 'description'
                }
            ]

            this.rows = this.permissions

            this.toolbarOptions = {
                densityButton: true,
                filterButton: true
            }

            this.tableOptions = {
                rowSelect: true
            }

            this.paginationOptions = {
                boundaryLinks: true,
                pageSelect: true,
                pageSizes: [10, 25, 50]
            }

            this.chipFilterOptions = {
                staticResults: this.staticResults
            }

            this.cardContent = '<permission-card permission="row" current-user="$ctrl.currentUser"></permission-card>'

            this.params = {
                page: this.listService.page - 1,
                pageSize: this.listService.pageSize,
                sort: this.listService.sort
            }

            this.getRows()
        }

        getRows() {
            let start = this.params.page * this.params.pageSize
            let end = start + this.params.pageSize
            this.rows = _.slice(this.permissions, start, end)
        }

        changePage(params) {
            this.params.page = params.page
            this.params.pageSize = params.pageSize

            this.listService.page = params.page + 1
            this.listService.pageSize = params.pageSize

            return this.getRows()
        }

        changeSort(params) {
            this.params.sort = params.sort

            this.listService.sort = params.sort

            return this.getRows()
        }

        changeDensity(params) {
            return this.listService.density = params.density
        }

        checkChipsCustom() {
            return !!this.chips
        }

        clearChipsCustom() {
            this.getRows()
            return _.remove(this.chips)
        }

        chipsUpdated(selectedChips) {
            if (_.get(selectedChips.selectedChips, 'length')) {
                let results = _.reduce(selectedChips.selectedChips, (result, chip) => {
                    let found = this.$filter('filter')(this.permissions, chip.term)
                    result.push(found)
                    return result
                }, [])

                //find the results that match all chips and put them first in the results
                //then put the rest of the results in the list - this should relatively
                //simulate how elastic search returns search results
                this.rows = _.uniq(_.intersection(...results).concat(_.flatten(results)))

                this.total = _.size(this.rows)
            } else {
                this.getRows()

                this.total = _.size(this.permissions)
            }

            //stop the state reload from happening
            return true
        }
    }
}
