// alk-maps/src/main/layers/VectorMarkerLayer.js
var LayerApi = require('./LayerApi')

module.exports = function(ALKMaps, map, options) {
    var layer = new ALKMaps.Layer.VectorMarkers(
        options.layerName || 'Vector Marker Layer'
    )
    layer.options = options
    map.addLayer(layer)

    var api = new LayerApi(ALKMaps, map, layer, null, options)
    api.eventsEnabled = true

    api.addItem = function(item) {
        var lonLat = api.createLonLat(item.location.lon, item.location.lat)

        var style = null
        if (options.markerType === 'custom') {
            style = {
                externalGraphic: options.externalGraphic,
                graphicHeight: options.graphicHeight || 26,
                graphicWidth: options.graphicWidth || 26,
                graphicXOffset: options.graphicXOffset || -13,
                graphicYOffset: options.graphicYOffset || -26
            }
        }

        var marker = layer.createMarker(
            lonLat, {
                markerType: options.markerType || 'green'
            },
            style)
        marker.attributes = item
    }

    api.clearItems = function() {
        layer.clearMarkers()
    }

    return api
}
